tests/src/compiler/hipClassKernel.cpp

```diff
@@ -0,0 +1,492 @@
+  result_ecd[tid] = (sizeof(testClassEmpty) == 1);
+}
+
+// check object addresses are not same
+__global__ void
+emptyClassKernelObj(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  testClassEmpty ob1, ob2;
+  result_ecd[tid] = (&ob1 != &ob2);
+}
+
+__global__ void
+sizeClassP1Kernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeP1)%6 == 0);
+}
+#ifdef ENABLE_DESTRUCTOR_TEST
+// check if constr and destr are accessible from kernel
+__global__ void
+testDeviceClassKernel() {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  testDeviceClass ob1;
+  testDeviceClass ob2;
+  ob2.iVar = 10;
+}
+#endif
+
+// tests for classes >8 bytes
+__global__ void
+sizeClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeA)%16 == 0);
+}
+
+__global__ void
+sizeDerivedClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeDerived)%24 == 0);
+}
+
+__global__ void
+sizeDerivedClassPragmaKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeDerived2)%20 == 0);
+}
+
+__global__ void
+sizeVirtualClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeDV)%16 == 0);
+}
+
+__global__ void
+sizeDerivedVirtualClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeDerivedDV)%16 == 0);
+}
+
+__global__ void
+sizeClassBKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeB)%12 == 0);
+}
+
+__global__ void
+sizeClassCKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeC)%16 == 0);
+}
+
+__global__ void
+sizeClassP2Kernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeP2)%13 == 0);
+}
+
+__global__ void
+sizeClassP3Kernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeP3)%8 == 0);
+}
+
+__global__ void
+sizeVirtDerClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeVirtDer)%24 == 0);
+}
+
+__global__ void
+sizeVirtDerPackClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeVirtDerPack)%24 == 0);
+}
+
+__global__ void
+sizeVirtDerMultiClassKernel(bool* result_ecd) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  result_ecd[tid] = (sizeof(testSizeDerMulti)%48 == 0);
+}
+
+int main() {
+  bool *result_ecd, *result_eob, *result_ech, *result_eoh;
+  size_t NBOOL = BLOCKS * sizeof(bool);
+  HIPCHECK(hipHostMalloc(reinterpret_cast<void**>(&result_ech),
```

Comment:
Many tests have similar steps. Can you write a template function for such tests and call the template function?

Reply:
Hi @yxsamliu , I have refactored the kernels and reduced the number to 3.